controllers/reconcilers/kserve_isvc_gateway_reconciler.go

```diff
@@ -0,0 +1,261 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reconcilers
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/go-logr/logr"
+	kservev1beta1 "github.com/kserve/kserve/pkg/apis/serving/v1beta1"
+	"github.com/opendatahub-io/odh-model-controller/controllers/comparators"
+	"github.com/opendatahub-io/odh-model-controller/controllers/constants"
+	"github.com/opendatahub-io/odh-model-controller/controllers/processors"
+	"github.com/opendatahub-io/odh-model-controller/controllers/resources"
+	"github.com/opendatahub-io/odh-model-controller/controllers/utils"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	istiov1beta1 "istio.io/api/networking/v1beta1"
+	istioclientv1beta1 "istio.io/client-go/pkg/apis/networking/v1beta1"
+)
+
+var _ SubResourceReconciler = (*KserveGatewayReconciler)(nil)
+var meshNamespace string
+var destSecretName string
+var portName string
+
+type KserveGatewayReconciler struct {
+	client         client.Client
+	secretHandler  resources.SecretHandler
+	gatewayHandler resources.GatewayHandler
+	deltaProcessor processors.DeltaProcessor
+}
+
+func NewKserveGatewayReconciler(client client.Client) *KserveGatewayReconciler {
+	return &KserveGatewayReconciler{
+		client:         client,
+		secretHandler:  resources.NewSecretHandler(client),
+		gatewayHandler: resources.NewGatewayHandler(client),
+		deltaProcessor: processors.NewDeltaProcessor(),
+	}
+}
+
+func (r *KserveGatewayReconciler) Reconcile(ctx context.Context, log logr.Logger, isvc *kservev1beta1.InferenceService) error {
+	log.V(1).Info("Reconciling KServe local gateway for Kserve InferenceService")
+
+	_, meshNamespace = utils.GetIstioControlPlaneName(ctx, r.client)
+	destSecretName = fmt.Sprintf("%s-%s", isvc.Name, isvc.Namespace)
+	portName = fmt.Sprintf("%s-%s", "https", isvc.Name)
+
+	// return if Address.URL is not set
+	if isvc.Status.Address != nil && isvc.Status.Address.URL == nil {
+		log.V(1).Info("Waiting for the URL as the InferenceService is not ready yet")
+		return nil
+	}
+
+	// return if serving cert secret in the source namespace is not created
+	srcCertSecret, err := r.secretHandler.Get(ctx, types.NamespacedName{Name: isvc.Name, Namespace: isvc.Namespace})
```

Comment:
I'm still unsure that this is really going to work. Units may be passing, because `suite_test.go` is not configuring for the client cache, like in `main.go`. But in the actual deployment, it should be always returning _not found_.

Reply:
The secretHandler just get secret with name and namespace. This is not working with cache stuff.
So I don't think it has something to do with cache.

Reply:
I used APIReader instead of clientset. It is easier to use in this case because manager has this object.